File: backend/app/schemas/reference_pack.py
"""参考包 API Schema

设计要点：
- 列表项 (ReferencePackSummary) 不返回维度 JSON 字段，避免列表加载过重
- 详情 (ReferencePackDetail) 一次返回全部维度内容；`pipeline_version` 区分老包（2，只读）与 V5 包（5）
- 维度内部结构使用 Dict[str, Any]，便于 prompt 演进时灵活扩展
- 项目挂载关联使用独立的 schema，包含默认引用配置

V5 维度集合（book_dissect_v5_design.md §5）：synopsis / bridges / style / character_archive / methodology / structure / corpus。
archetypes / worldbuilding / entities / relations / events 已随实体图谱抽取删除，老包这些字段不再下发。
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field, ConfigDict


# ============================================================
# 参考包列表 / 详情
# ============================================================


class ReferencePackSummary(BaseModel):
    """参考包列表项（不含 5 tab 详细内容，列表加载用）"""

    id: str
    user_id: str
    task_id: str
    source_book_title: str
    status: Literal["generating", "ready", "partial", "failed"]
    generated_dimensions: List[str] = Field(
        default_factory=list,
        description="已成功生成的维度（V5：synopsis/bridges/style/character_archive/methodology/structure）",
    )
    pipeline_version: int = Field(
        default=2,
        description="拆书流水线版本：2 = V2-V4 老包（只读，建议重新抽取）；5 = V5 拆书卡 / 情节单元 / 骨架",
    )
    error_message: Optional[str] = None
    attached_project_count: int = Field(
        default=0,
        description="已挂载到的项目数（便于用户判断是否仍在使用）",
    )
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class ReferencePackDetail(BaseModel):
    """参考包详情（含全部维度内容；详情页一次返回）"""

    id: str
    user_id: str
    task_id: str
    source_book_title: str
    status: Literal["generating", "ready", "partial", "failed"]
    generated_dimensions: List[str] = Field(default_factory=list)
    pipeline_version: int = Field(default=2, description="2 = 老包（只读）；5 = V5")
    error_message: Optional[str] = None

    # V5 六个维度（老包的同名字段是 V2-V4 形状，前端按 pipeline_version 分流渲染）
    synopsis: Optional[Dict[str, Any]] = Field(
        None,
        description=(
            "全书骨架（V5）：genre_tag / one_line_premise / main_conflict / golden_finger / stages / "
            "top_payoffs / growth_system / power_system / long_foreshadowing / reading_promise / opening_strategy"
        ),
    )
    bridges: Optional[Dict[str, Any]] = Field(
        None,
        description=(
            "桥段库聚合（V5）：arc_count / avg_arc_length / arc_length_distribution / payoff_type_distribution / "
            "payoff_density / typical_arcs / role_pattern；全部情节单元走 /book-dissect/{task_id}/arcs"
        ),
    )
    style: Optional[Dict[str, Any]] = Field(
        None,
        description="文风指纹（V5）：name / description / prompt_content / traits / dialogue_style / narration_habits / avoid_list / metrics / examples",
    )
    character_archive: Optional[Dict[str, Any]] = Field(
        None,
        description="人物功能谱（V5）：protagonist / allies / antagonists / function_slots",
    )
    methodology: Optional[Dict[str, Any]] = Field(
        None,
        description="写法手册：golden_finger_pattern / opening_hook_pattern / facepunch_rhythm / power_progression / highlight_density",
    )
    structure: Optional[Dict[str, Any]] = Field(
        None,
        description="结构统计（V5，无 LLM）：节奏 / 钩子 / 爽点密度 / 张力十分位 / 单元长度分布",
    )

    attached_project_count: int = 0
    created_at: datetime
    updated_at: Optional[datetime] = None


# ============================================================
# 项目挂载关联
# ============================================================


# 7 个引用维度（V5）：6 个参考包 JSON 维度 + corpus（拆书卡检索）
# 注意：必须与前端 @/frontend/src/types/reference_pack.ts 的 ReferenceDimension 保持一致，
# 否则 deep 档挂载会 422（请求校验）/500（响应校验）。
ReferenceDimension = Literal[
    "synopsis", "bridges", "style", "character_archive", "methodology", "structure",
    "corpus",
]
ReferenceStrength = Literal["light", "medium", "deep"]


class AttachReferencePackRequest(BaseModel):
    """挂载参考包到项目的请求 body"""

    pack_id: str = Field(..., description="要挂载的参考包 ID")
    default_dimensions: Optional[List[ReferenceDimension]] = Field(
        default=None,
        description="默认引用维度（一键仿写弹板的初始勾选状态）；省略则按 strength 推断",
    )
    default_strength: ReferenceStrength = Field(
        default="medium",
        description="默认参考强度：light(仅文风) / medium(文风+方法论) / deep(全维度)",
    )


class UpdateAttachmentRequest(BaseModel):
    """更新已挂载参考包的默认配置（PATCH）"""

    default_dimensions: Optional[List[ReferenceDimension]] = None
    default_strength: Optional[ReferenceStrength] = None


class ProjectReferencePackResponse(BaseModel):
    """项目已挂载参考包列表项（含来源参考包元信息）"""

    id: str = Field(..., description="挂载关联表主键")
    project_id: str
    pack_id: str
    pack_summary: ReferencePackSummary = Field(..., description="冗存参考包元信息便于前端展示")
    default_dimensions: List[ReferenceDimension] = Field(default_factory=list)
    default_strength: ReferenceStrength = "medium"
    attached_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ============================================================
# 操作响应
# ============================================================


class AttachReferencePackResponse(BaseModel):
    """挂载成功响应"""

    attachment_id: str
    project_id: str
    pack_id: str
    default_dimensions: List[ReferenceDimension]
    default_strength: ReferenceStrength

File: frontend/src/types/reference_pack.ts
/**
 * 拆书参考包类型定义（V5）
 *
 * 参见：@/agent-docs/features/book_dissect_v5_design.md §5-§6
 * 后端 schema：@/backend/app/schemas/reference_pack.py
 */

export type ReferencePackStatus = 'generating' | 'ready' | 'partial' | 'failed';

/**
 * 注入维度（与后端 ReferenceDimension Literal 一致，否则挂载请求会 422）
 */
export type ReferenceDimension =
  | 'synopsis' // 全书骨架（类型 / 前提 / 大矛盾 / 金手指 / 阶段 / 爽点）
  | 'bridges' // 桥段库（情节单元聚合 + 典型单元）
  | 'style' // 文风指纹（prompt_content + 量化指标 + 例句）
  | 'character_archive' // 人物功能谱
  | 'methodology' // 写法手册
  | 'structure' // 结构统计（钩子 / 节奏 / 爽点密度 / 张力曲线）
  | 'corpus'; // 拆书卡检索（按本次内容 BM25 命中原书章）

export type ReferenceStrength = 'light' | 'medium' | 'deep';

/** V5 流水线版本号；小于此值的参考包为 V2-V4 老包（只读，建议重新抽取） */
export const V5_PIPELINE_VERSION = 5;

export function isV5Pack(pack: { pipeline_version?: number | null } | null | undefined): boolean {
  return (pack?.pipeline_version ?? 2) >= V5_PIPELINE_VERSION;
}

/**
 * 参考包列表项（不含维度正文）
 */
export interface ReferencePackSummary {
  id: string;
  user_id: string;
  task_id: string;
  source_book_title: string;
  status: ReferencePackStatus;
  generated_dimensions: string[];
  /** 2 = V2-V4 老包；5 = V5 */
  pipeline_version: number;
  error_message: string | null;
  attached_project_count: number;
  created_at: string;
  updated_at: string | null;
}

/**
 * 参考包详情（六个维度正文一次返回）
 *
 * 维度内部为灵活的 dict；null 表示该维度未生成（partial 状态下常见）。
 * pipeline_version >= 5 时六个字段是下方 *Data 形状；老包同名字段是 V2-V4 形状，前端只读 JSON 展示。
 */
export interface ReferencePackDetail extends ReferencePackSummary {
  synopsis: Record<string, unknown> | null;
  bridges: Record<string, unknown> | null;
  style: Record<string, unknown> | null;
  character_archive: Record<string, unknown> | null;
  methodology: Record<string, unknown> | null;
  structure: Record<string, unknown> | null;
}

/**
 * 项目已挂载参考包列表项
 */
export interface ProjectReferencePackItem {
  id: string; // 关联表主键
  project_id: string;
  pack_id: string;
  pack_summary: ReferencePackSummary;
  default_dimensions: ReferenceDimension[];
  default_strength: ReferenceStrength;
  attached_at: string;
}

/**
 * 挂载参考包请求 body
 */
export interface AttachReferencePackRequest {
  pack_id: string;
  default_dimensions?: ReferenceDimension[];
  default_strength?: ReferenceStrength;
}

export interface UpdateAttachmentRequest {
  default_dimensions?: ReferenceDimension[];
  default_strength?: ReferenceStrength;
}

export interface AttachReferencePackResponse {
  attachment_id: string;
  project_id: string;
  pack_id: string;
  default_dimensions: ReferenceDimension[];
  default_strength: ReferenceStrength;
}

// ============================================================
// V5 六个维度的"软约束"类型（实际仍为 Record<string, unknown>，
// 此处仅作 IDE 自动补全提示，方便组件渲染时识别字段名）
// 后端产出：skeleton_builder / style_fingerprint / dissect_stats
// ============================================================

/** 全书骨架里的一个阶段（skeleton_builder.build_stages） */
export interface SkeletonStage {
  title?: string;
  core_conflict?: string;
  protagonist_goal?: string;
  key_upgrades?: string;
  signature_arc?: string;
  ending_hook?: string;
  arc_start?: number;
  arc_end?: number;
  chapter_start?: number;
  chapter_end?: number;
  status?: string;
  origin?: 'llm' | 'fallback' | string;
  [k: string]: unknown;
}

/** synopsis：全书骨架（skeleton_builder.build_skeleton） */
export interface SkeletonData {
  genre_tag?: string;
  one_line_premise?: string;
  main_conflict?: string;
  golden_finger?: { what?: string; how_it_works?: string; evolution?: string[]; [k: string]: unknown } | string | null;
  stages?: SkeletonStage[];
  top_payoffs?: Array<{ stage?: string; arcs?: string; buildup?: string; trigger?: string; reward?: string; [k: string]: unknown }>;
  growth_system?: string;
  power_system?: string;
  long_foreshadowing?: Array<{ setup?: string; payoff?: string; role?: string; [k: string]: unknown }>;
  reading_promise?: string;
  opening_strategy?: string;
  pipeline_version?: number;
  [k: string]: unknown;
}

/** 情节单元（v5_types.StoryArc；bridges.typical_arcs 与 GET /arcs 同形） */
export interface StoryArcData {
  arc_index: number;
  start_chapter: number;
  end_chapter: number;
  title?: string;
  function?: string;
  boundary_reason?: string;
  structure?: string;
  protagonist_chain?: string;
  emotion_curve?: string;
  payoff?: string;
  payoff_type?: string;
  golden_finger_usage?: string;
  character_changes?: string;
  gains_costs?: string;
  foreshadowing?: string;
  chapter_roles?: Record<string, string>;
  tension_peak_chapter?: number | null;
  origin?: 'llm' | 'fallback' | string;
  [k: string]: unknown;
}

/** bridges：桥段库聚合（dissect_stats.build_bridges_payload） */
export interface BridgesV5Data {
  pipeline_version?: number;
  arc_count?: number;
  avg_arc_length?: number;
  arc_length_distribution?: Record<string, number>;
  payoff_type_distribution?: Record<string, number>;
  payoff_density?: string;
  typical_arcs?: StoryArcData[];
  role_pattern?: { avg_intro_chapters?: number; avg_build_chapters?: number; payoff_position_ratio?: number; [k: string]: unknown };
  [k: string]: unknown;
}

/** style：文风指纹（style_fingerprint.StyleFingerprintBuilder.build） */
export interface StyleFingerprintData {
  name?: string;
  description?: string;
  prompt_content?: string;
  traits?: string[];
  dialogue_style?: string;
  narration_habits?: string;
  avoid_list?: string[];
  /** style_stats.compute_style_metrics 的量化指标 */
  metrics?: Record<string, unknown>;
  examples?: Array<{ kind?: string; chapter?: number; text?: string }>;
  pipeline_version?: number;
  [k: string]: unknown;
}

/** 老包 / 通用文风字段（导入写作风格库只用到这几项） */
export type StyleData = Pick<StyleFingerprintData, 'name' | 'description' | 'prompt_content' | 'traits'>;

/** character_archive：人物功能谱（skeleton_builder.build_character_functions） */
export interface CharacterFunctionsData {
  protagonist?: {
    name?: string;
    persona?: string;
    golden_finger?: string;
    flaws_and_pressure?: string;
    growth_track?: Array<{ stage?: string; state?: string }>;
    [k: string]: unknown;
  } | null;
  allies?: Array<{ name?: string; function_role?: string; arc_span?: string; technique?: string; [k: string]: unknown }>;
  antagonists?: Array<{ name?: string; tier?: string; conflict_nature?: string; escalation?: string; outcome?: string; [k: string]: unknown }>;
  function_slots?: Array<{ slot?: string; how_used?: string; [k: string]: unknown }>;
  pipeline_version?: number;
  [k: string]: unknown;
}

/** structure：结构统计（dissect_stats.build_structure_stats） */
export interface StructureStatsData {
  chapter_count?: number;
  avg_chapter_words?: number;
  pace_distribution?: Record<string, number>;
  tension_by_decile?: number[];
  hook_type_distribution?: Record<string, number>;
  hook_rate?: number;
  payoff_chapter_rate?: number;
  payoff_density_chapters?: number;
  function_tag_distribution?: Record<string, number>;
  function_tags_by_decile?: string[][];
  arc_count?: number;
  avg_arc_length?: number;
  arc_length_distribution?: Record<string, number>;
  payoff_type_distribution?: Record<string, number>;
  hook_examples?: Array<{ chapter?: number; type?: string; text?: string }>;
  [k: string]: unknown;
}

/** methodology：写法手册（V3 五键形状未变） */
export interface MethodologyData {
  golden_finger_pattern?: {
    type?: string;
    balance_mechanism?: string;
    evolution_pattern?: string;
    writing_tips?: string;
    [k: string]: unknown;
  } | null;
  opening_hook_pattern?: {
    hook_type?: string;
    first_chapter_strategy?: string;
    writing_tips?: string;
    [k: string]: unknown;
  } | null;
  facepunch_rhythm?: {
    small_facepunch_freq?: string;
    big_facepunch_freq?: string;
    three_elements_pattern?: string;
    writing_tips?: string;
    [k: string]: unknown;
  } | null;
  power_progression?: {
    system_type?: string;
    level_count?: number;
    pace?: string;
    writing_tips?: string;
    [k: string]: unknown;
  } | null;
  highlight_density?: {
    small_per_n_chapters?: number;
    medium_per_n_chapters?: number;
    big_per_n_chapters?: number;
    writing_tips?: string;
    [k: string]: unknown;
  } | null;
}

// ============================================================
// 一键仿写（V3 R5）
// 后端 schema：@/backend/app/schemas/imitation.py
// ============================================================

/**
 * 一键仿写请求体
 *
 * pack_ids / dimensions / strength 全部可省略：
 * - 省略 pack_ids → 使用项目所有已挂载且 ready 的参考包
 * - 省略 dimensions → 取所选 pack 挂载关联的 default_dimensions 并集
 * - 省略 strength → 取所选 pack 中"最深"者
 */
export interface ImitateChapterRequest {
  user_intent: string;
  target_chapter_id?: string;
  pack_ids?: string[];
  dimensions?: ReferenceDimension[];
  strength?: ReferenceStrength;
  target_word_count?: number;
  style_id?: number;
}

export interface ImitationPackUsage {
  pack_id: string;
  source_book_title: string;
  dimensions: string[];
}

export interface ImitatePromptPreview {
  system_prompt: string;
  user_prompt: string;
  used_packs: ImitationPackUsage[];
  used_dimensions: string[];
  strength: ReferenceStrength;
  target_word_count: number;
  project_context_chars: number;
  reference_chars: number;
  extras?: Record<string, unknown>;
}

/** SSE 流式事件载荷（meta 事件） */

